sqlmesh/dbt/adapter.py

```diff
@@ -98,19 +98,49 @@ def quote(self, identifier: str) -> str:
     def dispatch(self, name: str, package: t.Optional[str] = None) -> t.Callable:
         """Returns a dialect-specific version of a macro with the given name."""
         target_type = self.jinja_globals["target"]["type"]
-        references_to_try = [
-            MacroReference(package=f"{package}_{target_type}", name=f"{target_type}__{name}"),
-            MacroReference(package=package, name=f"{target_type}__{name}"),
-            MacroReference(package=package, name=f"default__{name}"),
+        macro_suffix = f"__{name}"
+
+        def _relevance(package_name_pair: t.Tuple[t.Optional[str], str]) -> t.Tuple[int, int]:
+            """Lower scores more relevant."""
+            macro_package, macro_name = package_name_pair
+
+            package_score = 0 if macro_package == package else 1
+            name_score = 1
+
+            if macro_name.startswith("default"):
+                name_score = 2
+            elif macro_name.startswith(target_type):
+                name_score = 0
+
+            return name_score, package_score
+
+        jinja_env = self.jinja_macros.build_environment(**self.jinja_globals).globals
+        packages_to_check: t.List[t.Optional[str]] = [
+            package,
+            *(k for k in jinja_env if k.startswith("dbt")),
         ]
+        candidates = {}
+        for macro_package in packages_to_check:
+            macros: t.Dict[str, t.Any] = (
+                jinja_env.get(macro_package, {}) if macro_package else jinja_env  # type: ignore
+            )
+            matching_macro_names = [k for k in macros if k.endswith(macro_suffix)]
+            candidates.update(
+                {
+                    (macro_package, macro_name): macros[macro_name]
+                    for macro_name in matching_macro_names
```

Comment:
Is it possible to do the matching macro names work here instead of its own loop? 

```
(macro_package, macro_name): macros[macro_name]
for macro_name in macros
if macro_name.endswith(macro_suffix)
```

Reply:
ah, good point

Reply:
Fixed this